pow/src/main/java/tech/pegasys/teku/pow/fallback/FallbackAwareEth1Provider.java

```diff
@@ -0,0 +1,109 @@
+
+public class FallbackAwareEth1Provider implements Eth1Provider {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Eth1ProviderSelector eth1ProviderSelector;
+
+  public FallbackAwareEth1Provider(final Eth1ProviderSelector eth1ProviderSelector) {
+    this.eth1ProviderSelector = eth1ProviderSelector;
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1Block(final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockNumber));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1BlockWithRetry(
+      final UInt64 blockNumber, final Duration retryDelay, final int maxRetries) {
+    return run(
+        eth1Provider -> eth1Provider.getEth1BlockWithRetry(blockNumber, retryDelay, maxRetries));
+  }
+
+  @Override
+  public SafeFuture<Block> getGuaranteedEth1Block(final String blockHash) {
+    return run(eth1Provider -> eth1Provider.getGuaranteedEth1Block(blockHash));
+  }
+
+  @Override
+  public SafeFuture<Block> getGuaranteedEth1Block(final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.getGuaranteedEth1Block(blockNumber));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1Block(final String blockHash) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockHash));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1BlockWithRetry(
+      final String blockHash, final Duration retryDelay, final int maxRetries) {
+    return run(
+        eth1Provider -> eth1Provider.getEth1BlockWithRetry(blockHash, retryDelay, maxRetries));
+  }
+
+  @Override
+  public SafeFuture<Block> getLatestEth1Block() {
+    return run(Eth1Provider::getLatestEth1Block);
+  }
+
+  @Override
+  public SafeFuture<EthCall> ethCall(
+      final String from, final String to, final String data, final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.ethCall(from, to, data, blockNumber));
+  }
+
+  @Override
+  public SafeFuture<BigInteger> getChainId() {
+    return run(Eth1Provider::getChainId);
+  }
+
+  private <T> SafeFuture<T> run(final Function<Eth1Provider, SafeFuture<T>> task) {
+    return run(task, eth1ProviderSelector.candidateCount());
+  }
+
+  private <T> SafeFuture<T> run(
+      final Function<Eth1Provider, SafeFuture<T>> task, final int maxRetries) {
+    return SafeFuture.of(
+        () ->
+            task.apply(eth1ProviderSelector.bestCandidate())
+                .exceptionallyCompose(
+                    err -> {
+                      LOG.warn(
+                          "Error caught while calling Eth1Provider, switching provider for next calls",
+                          err);
+                      eth1ProviderSelector.updateBestCandidate();
+                      if (maxRetries > 0) {
+                        return run(task, maxRetries - 1);
+                      } else {
+                        return SafeFuture.failedFuture(err);
+                      }
+                    }));
```

Comment:
There's a race condition here where if two requests are in progress at the same time, they'll both be updating the best candidate and so wind up potentially iterating through the list multiple times.  I think it works out ok but the behaviour is pretty hard to reason about.

Reply:
Ok, i am not using an iterator as you suggested.